noir-projects/aztec-nr/aztec/src/utils/bytes.nr

```diff
@@ -1,88 +1,319 @@
+    let byte_idx = bytes.len() - 1;
+    let mut byte = bytes[byte_idx];
+    for j in 0..N - ((N / 8) * 8) {
+        let bit_idx = (byte_idx) * 8 + j;
+        bits[bit_idx] = (byte & 1) as u1;
+        byte >>= 1;
+    }
+
+    bits
+}
+
+/// Returns random bits, in chunks, according to the provided numeric generics.
+unconstrained fn get_chunks_of_random_bits<let NUM_CHUNKS: u32, let NUM_BITS: u32>() -> [[u1; NUM_BITS]; NUM_CHUNKS] {
+    let random_bits = get_random_bits::<NUM_CHUNKS * NUM_BITS>();
+    let mut chunks: [[u1; NUM_BITS]; NUM_CHUNKS] = [[0; NUM_BITS]; NUM_CHUNKS];
+    for i in 0..NUM_CHUNKS {
+        let mut chunk: [u1; NUM_BITS] = [0; NUM_BITS];
+        for j in 0..NUM_BITS {
+            chunk[j] = random_bits[i * NUM_BITS + j];
+        }
+        chunks[i] = chunk;
+    }
+    chunks
+}
+
+/**
+ * Helper fn for be/le_bytes_to_padded_fields() fns.
+ */
+fn pad_31_byte_fields_with_random_bits<let N: u32>(input: [Field; N]) -> [Field; N] {
+    /// Safety: we can't constrain whether or not the returned bits are actually random;
+    /// we can only rely on it being in the caller's best interest.
+    let chunks_of_random_bits = unsafe { get_chunks_of_random_bits::<((N + 30) / 31), 6>() };
+    let mut fields = input;
+    // We push the random bits to the "big end" of each field
+    let TWO_POW_248 = 0x100000000000000000000000000000000000000000000000000000000000000;
+    for i in 0..fields.len() {
+        let mut shift = TWO_POW_248;
+        for j in 0..5 {
+            fields[i] += shift * chunks_of_random_bits[i][j] as Field;
+            shift *= 2;
+        }
+
+        // This might overflow the field modulus, so we need to be careful.
+        // By this point, shift = 2 ^ 253, so will populate the 254th bit (counting from 1)
+        let field_with_254th_bit = fields[i] + shift * chunks_of_random_bits[i][6 - 1] as Field;
+
+        // Expensive comparison (one for each field), but it's important -- if we want to hide that we were
+        // working with bytes -- that we include the 254th random bit when it fits.
+        fields[i] = if (fields[i].lt(field_with_254th_bit)) {
+            field_with_254th_bit
+        } else {
+            fields[i]
+        };
+    }
+    fields
+}
+
+/**
+ * The above functions be_bytes_31_to_fields and le_bytes_31_to_fields have a privacy-leakage problem:
+ * they result in fields which are all 248-bits instead of 254-bits.
+ * That means if the above functions are used to convert a ciphertext from bytes to
+ * fields, it leaks (with very high probability) to all observers of the log that the
+ * ciphertext was created with a "bit twiddly" encryption scheme like AES.
+ * This reduces privacy sets for all apps.
+ *
+ * This function seeks to "pad" those remaining 6 bits with randomness, so that all
+ * encrypted logs look even more similar (full fields of 254-bits of random-looking
+ * information). Of course, this adds some extra constraints, unfortunately.
+ */
```

Comment:
Did this exist before this PR?

Reply:
Yep, it sure did! It was one of the reasons I delved into the bytes.nr file and began refactoring it.

Reply:
I didn't go as far as plumbing this function into the log computation yet.